[SPIR-V] Redesign symbolic operands (SPIRVEnums.h) in TableGen

michalpaszkowski opened the pull request:
1. Move symbolic/named operands enum definitions from macros to TableGen
2. Reimplement helper functions for getting versioning/capabilities/
extensions to use data from TableGen tables.
3. Generate instruction printing methods for symbolic/named operands
from TableGen

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/KhronosGroup/LLVM-SPIRV-Backend?pullRequest=134) <br/>All committers have signed the CLA.

iliya-diyachkov commented on llvm/lib/Target/SPIRV/SPIRVOperands.td:
```diff
@@ -0,0 +1,1515 @@
+//defm ImageRect : CapabilityOperand<36, 0, 0, [], [SampledRect]>;
+defm SampledRect : CapabilityOperand<37, 0, 0, [], [Shader]>;
+
+defm ImageRect : CapabilityOperand<36, 0, 0, [], [SampledRect]>;
```
Shouldn't we just uncomment this on line 301?

michalpaszkowski replied:
Right, thanks! Forgot about these lines! The commented lines should be removed - I will update the patch. The definitions for _ImageRect_ must be lower (at line 304) since _ImageRect_ requires _SampledRect_ capability, which is defined just above at line 302. For the very same reason _ImageCubeArray_ is defined lower.

I don't think there is any way to "forward declare" TableGen entries -- that would resolve this problem. Though this issue only happens for the two _ImageRect_ and _ImageCubeArray_ definitions.

iliya-diyachkov commented on llvm/lib/Target/SPIRV/SPIRVOperands.td:
```diff
@@ -0,0 +1,1515 @@
+defm Image1D : CapabilityOperand<44, 0, 0, [], [Sampled1D]>;
+defm SampledCubeArray : CapabilityOperand<45, 0, 0, [], [Shader]>;
+
+defm ImageCubeArray : CapabilityOperand<34, 0, 0, [], [SampledCubeArray]>;
```
The same question: shouldn't we just uncomment this on line 299?

iliya-diyachkov commented:
In general it looks fine, great job, Michal!

By the way, the patch contains whitespace errors, please fix:
```
> git apply patchFromMichal
patchFromMichal:2141: trailing whitespace.
// This file contains TableGen generated enum definitions, mnemonic lookup
patchFromMichal:2399: trailing whitespace.
// Lookup table for matching symbolic operands (category + 32-bit value) to
patchFromMichal:2423: trailing whitespace.
// Lookup table for matching symbolic operands (category + 32-bit value) to
patchFromMichal:2443: trailing whitespace.
  // Function for looking up a (the first) capability by category + value. Next
patchFromMichal:2461: trailing whitespace.

warning: squelched 108 whitespace errors
warning: 113 lines add whitespace errors.
```

michalpaszkowski commented:
> In general it looks fine, great job, Michal!
> 
> By the way, the patch contains whitespace errors, please fix:
> 
> ```
> > git apply patchFromMichal
> patchFromMichal:2141: trailing whitespace.
> // This file contains TableGen generated enum definitions, mnemonic lookup
> patchFromMichal:2399: trailing whitespace.
> // Lookup table for matching symbolic operands (category + 32-bit value) to
> patchFromMichal:2423: trailing whitespace.
> // Lookup table for matching symbolic operands (category + 32-bit value) to
> patchFromMichal:2443: trailing whitespace.
>   // Function for looking up a (the first) capability by category + value. Next
> patchFromMichal:2461: trailing whitespace.
> 
> warning: squelched 108 whitespace errors
> warning: 113 lines add whitespace errors.
> ```

Thank you @iliya-diyachkov! Will fix!

zuban32 commented on llvm/lib/Target/SPIRV/SPIRVOperands.cpp:
```diff
@@ -0,0 +1,178 @@
+#include "SPIRVGenTables.inc"
+} // end anonymous namespace
+
+std::string
```
I think llvm::StringRef would be more appropriate here and in all the similar usecases below

zuban32 commented on llvm/lib/Target/SPIRV/SPIRVOperands.cpp:
```diff
@@ -0,0 +1,178 @@
+std::string
+getSymbolicOperandMnemonic(::OperandCategory::OperandCategory Category,
+                           int32_t Value) {
+  const SymbolicOperand *lookup =
```
Codestyle - s/lookup/Lookup

zuban32 commented on llvm/lib/Target/SPIRV/SPIRVOperands.cpp:
```diff
@@ -0,0 +1,178 @@
+}
+
+// TODO: Remove function after new SPIRVOpenCLBIFs is merged
+bool getSpirvBuilInIdByName(llvm::StringRef Name, BuiltIn::BuiltIn &BI) {
```
typo in the method name

zuban32 commented on llvm/lib/Target/SPIRV/SPIRVOperands.cpp:
```diff
@@ -0,0 +1,178 @@
+      ::OperandCategory::BuiltInOperand, BuiltInValue);
+
+  if (lookup)
+    return "__spirv_BuiltIn" + lookup->Mnemonic.str();
```
I'd say we need to have this string constant unique somewhere to be used by both OCLBIFs and here.

And I'm strongly against this '15' constant below, won't hurt using this constant string length instead

zuban32 commented on llvm/lib/Target/SPIRV/SPIRVOperands.td:
```diff
@@ -0,0 +1,1515 @@
+// required extension and capabilities.
+//===----------------------------------------------------------------------===//
+
+multiclass SymbolicOperandWithRequirments<OperandCategory category, bits<32> value, string mnemonic, bits<32> minVersion, bits<32> maxVersion, list<Extension> reqExtensions, list<Capability> reqCapabilities> {
```
typo s/SymbolicOperandWithRequirments/SymbolicOperandWithRequirements

zuban32 commented on llvm/lib/Target/SPIRV/SPIRVOperands.h:
```diff
@@ -0,0 +1,204 @@
+//===-- SPIRVOperands.h -  Top level definitions for SPIRV ------*- C++ -*-===//
```
IMO this file name is misleading, I'd suggest something like 'SPIRVSymbolicOperands' or something like that

zuban32 commented:
Other than a StringRef one all my comments are about minor or codestyle-related issues, so I agree to Iliya this looks good and fixes this terrible code we had before, so thank you Michal!

iliya-diyachkov commented on llvm/lib/Target/SPIRV/MCTargetDesc/SPIRVInstPrinter.cpp:
```diff
@@ -312,48 +320,25 @@ void SPIRVInstPrinter::printStringImm(const MCInst *MI, unsigned OpNo,
 }
 
-void SPIRVInstPrinter::printExtInst(const MCInst *MI, unsigned OpNo,
+void SPIRVInstPrinter::printExtension(const MCInst *MI, unsigned OpNo,
```
Please use `clang-format  -i -style=LLVM llvm/lib/Target/SPIRV/MCTargetDesc/SPIRVInstPrinter.cpp`

michalpaszkowski committed:
[SPIR-V] Redesign symbolic operands (SPIRVEnums.h) in TableGen

1. Move symbolic/named operands enum definitions from macros to TableGen
2. Reimplement helper functions for getting versioning/capabilities/
extensions to use data from TableGen tables.
3. Generate instruction printing methods for symbolic/named operands
from TableGen
